server/src/main/java/org/candlepin/controller/ContentManager.java

```diff
@@ -525,51 +527,147 @@ public Content updateContent(ContentData update, Owner owner, boolean regenerate
+        boolean regenerateEntitlementCerts) {
 
-        this.ownerContentCurator.removeOwnerFromContent(existing, owner);
-        if (this.ownerContentCurator.getOwnerCount(existing) == 0) {
-            // No one is using this product anymore; delete the entity (should clean up everything)
-            this.contentCurator.delete(existing);
+        if (owner == null) {
+            throw new IllegalArgumentException("owner is null");
         }
-        else {
-            // Clean up any dangling references to content
-            this.ownerContentCurator.removeOwnerContentReferences(existing, owner);
+
+        if (contentUuids != null && !contentUuids.isEmpty()) {
+            log.debug("Deleting content with UUIDs: {}", contentUuids);
+
+            List<Product> affectedProducts = this.productCurator
+                .getProductsByContentUuids(owner, contentUuids)
+                .list();
+
+            if (!affectedProducts.isEmpty()) {
+                log.debug("Updating {} affected products", affectedProducts.size());
+
+                if (!(contentUuids instanceof Set)) {
+                    // Convert this to a set so our filtering lookups aren't painfully slow
+                    contentUuids = new HashSet<String>(contentUuids);
+                }
+
+                // Get the collection of content those products use, throwing out the ones we'll be
+                // removing shortly
+                Map<String, Content> affectedProductsContent = new HashMap<String, Content>();
+                for (Content content : this.contentCurator.getContentByProducts(affectedProducts)) {
+                    if (!contentUuids.contains(content.getUuid())) {
+                        affectedProductsContent.put(content.getId(), content);
+                    }
+                }
+
+                // Convert our affectedProducts into DTOs (hoping Hibernate uses its entity cache
+                // instead of pulling down the content list for each product...)
+                Map<String, ProductData> affectedProductData = new HashMap<String, ProductData>();
+                for (Product product : affectedProducts) {
+                    ProductData pdata = product.toDTO();
+
+                    Iterator<ProductContentData> pcd = pdata.getProductContent().iterator();
+                    while (pcd.hasNext()) {
+                        ContentData cdata = pcd.next().getContent();
+
+                        if (!affectedProductsContent.containsKey(cdata.getId())) {
+                            pcd.remove();
+                        }
+                    }
+
+                    affectedProductData.put(pdata.getId(), pdata);
+                }
+
+                // Perform a micro-import for these products using the content map we just built
+                log.debug("Performing micro-import for products: {}", affectedProductData);
+                this.productManager.importProducts(owner, affectedProductData, affectedProductsContent);
```

Comment:
Originally we just inserted the data directly via updateProduct/content. Using the import methods allows us to make use of the batching optimizations made for import process here. I called it a "micro-import" because it's not doing the full import process, nor will the data be as large as a typical refresh/import.